From 9b16d2c410ef545ba9d587f5c0c878165266c59e Mon Sep 17 00:00:00 2001
From: Suniel Mahesh <sunil@amarulasolutions.com>
Date: Thu, 2 Jul 2020 16:08:54 +0530
Subject: [PATCH 1/2] arm64: rockchip: rock-pi-n10: Enable AP6398S WiFi/BT
 support

Radxa has a custom ROCK Pi M.2 Wireless Module for rock-pi-n10
target which sits on an m.2 slot. This wireless module has a
AP6398S WiFi/BT chipset which is connected to target via SDIO
and UART. Enable AP6398S WiFi/BT on rock-pi-n10.

Signed-off-by: Suniel Mahesh <sunil@amarulasolutions.com>
---
 .../boot/dts/rockchip/rk3399pro-rock-pi-n10.dts    | 109 +++++++++++++++++++++
 1 file changed, 109 insertions(+)

diff --git a/arch/arm64/boot/dts/rockchip/rk3399pro-rock-pi-n10.dts b/arch/arm64/boot/dts/rockchip/rk3399pro-rock-pi-n10.dts
index a1783e7..9528998 100644
--- a/arch/arm64/boot/dts/rockchip/rk3399pro-rock-pi-n10.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399pro-rock-pi-n10.dts
@@ -15,4 +15,113 @@
 	model = "Radxa ROCK Pi N10";
 	compatible = "radxa,rockpi-n10", "vamrs,rk3399pro-vmarc-som",
 		     "rockchip,rk3399pro";
+
+	sdio_pwrseq: sdio-pwrseq {
+                status = "okay";
+                compatible = "mmc-pwrseq-simple";
+                clocks = <&hym8563>;
+                clock-names = "ext_clock";
+                pinctrl-names = "default";
+                pinctrl-0 = <&wifi_enable_h>;
+
+                /*
+                 * On the module itself this is one of these (depending
+                 * on the actual card populated):
+                 * - SDIO_RESET_L_WL_REG_ON
+                 * - PDN (power down when low)
+                 */
+                reset-gpios = <&gpio2 RK_PD3 GPIO_ACTIVE_LOW>;
+        };
+};
+
+&sdio0 {
+        #address-cells = <1>;
+        #size-cells = <0>;
+        bus-width = <4>;
+        clock-frequency = <50000000>;
+        cap-sdio-irq;
+        cap-sd-highspeed;
+        keep-power-in-suspend;
+        mmc-pwrseq = <&sdio_pwrseq>;
+        non-removable;
+        pinctrl-names = "default";
+        pinctrl-0 = <&sdio0_bus4 &sdio0_cmd &sdio0_clk>;
+        sd-uhs-sdr104;
+        status = "okay";
+
+        brcmf: wifi@1 {
+                compatible = "brcm,bcm4329-fmac";
+                reg = <1>;
+                interrupt-parent = <&gpio0>;
+                interrupts = <RK_PA3 GPIO_ACTIVE_HIGH>;
+                interrupt-names = "host-wake";
+                pinctrl-names = "default";
+                pinctrl-0 = <&wifi_host_wake_l>;
+        };
+};
+
+&uart0 {
+        pinctrl-names = "default";
+        pinctrl-0 = <&uart0_xfer &uart0_cts &uart0_rts>;
+        status = "okay";
+
+        bluetooth {
+                compatible = "brcm,bcm43438-bt";
+                clocks = <&hym8563>;
+                clock-names = "ext_clock";
+                device-wakeup-gpios = <&gpio2 RK_PD2 GPIO_ACTIVE_HIGH>;
+		host-wakeup-gpios = <&gpio0 RK_PA5 GPIO_ACTIVE_HIGH>;
+                shutdown-gpios = <&gpio2 RK_PD4 GPIO_ACTIVE_HIGH>;
+                pinctrl-names = "default";
+                pinctrl-0 = <&bt_host_wake_l &bt_wake_l &bt_enable_h>;
+        };
+};
+
+&pinctrl {
+	sdio-pwrseq {
+                wifi_enable_h: wifi-enable-h {
+                        rockchip,pins =
+                                <2 RK_PD3 0 &pcfg_pull_none>;
+                };
+        };
+
+	sdio0 {
+                sdio0_bus4: sdio0-bus4 {
+                        rockchip,pins =
+                                <2 RK_PC4 1 &pcfg_pull_up_20ma>,
+                                <2 RK_PC5 1 &pcfg_pull_up_20ma>,
+                                <2 RK_PC6 1 &pcfg_pull_up_20ma>,
+                                <2 RK_PC7 1 &pcfg_pull_up_20ma>;
+                };
+
+                sdio0_cmd: sdio0-cmd {
+                        rockchip,pins =
+                                <2 RK_PD0 1 &pcfg_pull_up_20ma>;
+                };
+
+		sdio0_clk: sdio0-clk {
+                        rockchip,pins =
+                                <2 RK_PD1 1 &pcfg_pull_none_20ma>;
+		};
+	};
+
+	wifi {
+                wifi_host_wake_l: wifi-host-wake-l {
+                        rockchip,pins = <0 RK_PA3 RK_FUNC_GPIO &pcfg_pull_none>;
+                };
+        };
+
+	bt {
+                bt_enable_h: bt-enable-h {
+                        rockchip,pins = <2 RK_PD4 RK_FUNC_GPIO &pcfg_pull_none>;
+                };
+
+                bt_host_wake_l: bt-host-wake-l {
+                        rockchip,pins = <0 RK_PA5 RK_FUNC_GPIO &pcfg_pull_none>;
+                };
+
+                bt_wake_l: bt-wake-l {
+                        rockchip,pins = <2 RK_PD2 RK_FUNC_GPIO &pcfg_pull_none>;
+                };
+        };
 };
-- 
2.7.4

